[8.2] CMS Installer Fails During Repository Installation Due to Invalid XML Comment ("--" Not Permitted Within Comments)

**Description:** The Percussion CMS installation fails during the repository installation phase. The Apache Ant build terminates with an XML parsing error indicating that the installer encountered an invalid XML comment containing the string --.  **Steps to Reproduce:**  1. Launch the Percussion CMS installer. 2. Proceed with the installation using the default or desired configuration. 3. Allow the installer to execute the repository installation tasks. 4. Observe that the installation fails before completion.   **Actual Result:** The installation fails with the following error:  **BUILD FAILED**  ``` C:\Users\VIJAYA~1.BOD\AppData\Local\Temp\percInstallTmp_12973903112453842123\rxconfig\Installer\install.xml:645: The following error occurred while executing this line: C:\Users\VIJAYA~1.BOD\AppData\Local\Temp\percInstallTmp_12973903112453842123\rxconfig\Installer\install.xml:672: The following error occurred while executing this line: C:\Users\VIJAYA~1.BOD\AppData\Local\Temp\percInstallTmp_12973903112453842123\rxconfig\Installer\install.xml:279: The following error occurred while executing this line: C:\Users\VIJAYA~1.BOD\AppData\Local\Temp\percInstallTmp_12973903112453842123\rxconfig\Installer\installRepository.xml:205: The string "--" is not permitted within comments. ```  **Expected Result:** The CMS installer should complete successfully without XML parsing errors. Any XML files used by the installer should conform to the XML specification and must not contain invalid comment syntax.  **Impact:**  - CMS installation cannot be completed. - New installations are blocked until the invalid XML comment is corrected. - Users must manually investigate installer resources, preventing successful deployment.  **Possible Cause:**  - installRepository.xml contains an XML comment with the sequence --, which is invalid according to the XML specification. The installer should either: - Remove or correct the invalid comment, or - Validate installer XML files before packaging to prevent malformed XML from being distributed. 

 ## Agent progress (night-issue-prs)

| Slice | Issue | Status | PR | Notes | Updated |
|-------|-------|--------|----|-------|---------|
| Full fix: illegal `--` in Installer XML comments | #2290 | pr_opened | https://github.com/intersoftdatalabs-in/percussioncms/pull/2304 | Comment rewords + InstallerXmlWellFormedTest; live install smoke optional ops | 2026-08-07 16:49 UTC |

--- a/docs/ai-generated/code-reviews/issue-2290-installer-xml-comment-double-dash-erlang.md
+++ b/docs/ai-generated/code-reviews/issue-2290-installer-xml-comment-double-dash-erlang.md
@@ -1,0 +1,43 @@
+# Erlang review: fix/issue-2290-installer-xml-comment-double-dash
+
+**Date:** 2026-08-07  
+**Branch:** fix/issue-2290-installer-xml-comment-double-dash  
+**Base:** origin/main  
+**Reviewer persona:** Erlang (strict gate)  
+**Author agent:** Grok Build / main / grok-4.5
+
+## Summary
+
+Comment-only fix for invalid XML `--` sequences inside Installer Ant comments that abort CMS install (issue #2290). Adds a cheap well-formedness + comment-body regression test for top-level Installer Ant XML.
+
+## Scope
+
+- `modules/perc-distribution-tree/src/main/resources/distribution/rxconfig/Installer/installRepository.xml` — reword `--demo-sites` in comment
+- `modules/perc-distribution-tree/src/main/resources/distribution/rxconfig/Installer/remove_PercussionInstallation.xml` — reword `--clean-install-dir` in comment
+- `modules/perc-distribution-tree/src/test/java/com/percussion/distribution/install/InstallerXmlWellFormedTest.java` — new unit tests
+
+Cross-platform path review: test uses `Path.of` / `Files.list` / `Files.readString`; no hardcoded `/` or `\` joins beyond Path segments; module-relative paths match peer packaging tests. Clean.
+
+Memory patterns: none specific hit; installer packaging tests peer pattern followed.
+
+## Recommendation
+
+**approve**
+
+## Gate
+
+**May commit/push: yes**
+
+## Issues
+
+(none)
+
+## Evidence
+
+- Full-tree comment scan of Installer `*.xml`: only the two reworded sites had illegal `--`; post-fix count 0
+- `cd modules/perc-distribution-tree && ../../mvnw clean install` green
+- `InstallerXmlWellFormedTest`: Tests run: 2, Failures: 0, Errors: 0
+
+## Residual (ops / optional)
+
+Live install smoke remains optional human/ops follow-up; not required for this PR.

--- a/modules/perc-distribution-tree/src/main/resources/distribution/rxconfig/Installer/installRepository.xml
+++ b/modules/perc-distribution-tree/src/main/resources/distribution/rxconfig/Installer/installRepository.xml
@@ -202,7 +202,7 @@
 
 		<!-- Optional sample sites (Corporate Investments / Enterprise Investments).
 			Driven by the Java preinstall's demo-sites flag, propagated to ANT as -Dinstall.demo.sites=true
-			from Phase-1 options (wizard Yes / --demo-sites) via Main.execJar (#2192).
+			from Phase-1 options (wizard Yes / demo-sites CLI flag) via Main.execJar (#2192).
 			When true on new installs and upgrades: antcall installSampleSites chains an
 			RxffTableData/RxffTableDef PSTableAction pass that seeds the two sample sites plus
 			their ACLs, content types, nav, etc. RXLOCALE / RXLOCALEFORMAT are protected by

--- a/modules/perc-distribution-tree/src/main/resources/distribution/rxconfig/Installer/remove_PercussionInstallation.xml
+++ b/modules/perc-distribution-tree/src/main/resources/distribution/rxconfig/Installer/remove_PercussionInstallation.xml
@@ -3,9 +3,9 @@
 	<!-- Historical ANT stub for <InstallDir>/_Percussion_Installation. As of
 		issue #1157, obsolete install-root cleanup (including this path and PreInstall)
 		is owned by the preinstall Java helper com.percussion.preinstall.ObsoleteInstallDirCleaner,
-		invoked early on upgrade from Main when the operator confirms or passes --clean-install-dir.
-		This target remains a no-op so any antcall to this file does not double-delete
-		or fail. -->
+		invoked early on upgrade from Main when the operator confirms or passes the
+		clean-install-dir CLI flag. This target remains a no-op so any antcall to this
+		file does not double-delete or fail. -->
 	<target name="remove_PercussionInstallation">
 		<echo>Obsolete _Percussion_Installation cleanup is handled by
 			preinstall ObsoleteInstallDirCleaner (issue #1157); ANT target is a

--- a/modules/perc-distribution-tree/src/test/java/com/percussion/distribution/install/InstallerXmlWellFormedTest.java
+++ b/modules/perc-distribution-tree/src/test/java/com/percussion/distribution/install/InstallerXmlWellFormedTest.java
@@ -1,0 +1,122 @@
+/*
+ * Copyright (c) 2026 Intersoft Data Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.percussion.distribution.install;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.xml.sax.SAXException;
+
+/**
+ * Issue #2290: installer Ant XML must be well-formed. XML forbids the sequence {@code --} inside
+ * comments; a CLI flag mention such as {@code --demo-sites} inside {@code <!-- ... -->} causes
+ * Apache Ant to abort with {@code The string "--" is not permitted within comments} mid-install.
+ *
+ * <p>Cheap regression guard: parse each top-level Installer Ant script and scan comment bodies for
+ * consecutive hyphens. Nested {@code data/} table XML is covered by existing packaging tests.
+ */
+@Tag("UnitTest")
+class InstallerXmlWellFormedTest {
+
+  private static final Path INSTALLER_DIR =
+      Path.of("src", "main", "resources", "distribution", "rxconfig", "Installer");
+
+  /** Match XML comments (non-greedy, DOTALL). */
+  private static final Pattern XML_COMMENT = Pattern.compile("<!--(.*?)-->", Pattern.DOTALL);
+
+  @Test
+  @DisplayName("Installer Ant XML comments must not embed consecutive hyphens (#2290)")
+  void installerAntXmlCommentsHaveNoDoubleHyphen() throws IOException {
+    List<Path> scripts = listTopLevelInstallerXml();
+    assertFalse(scripts.isEmpty(), "expected Installer Ant XML under " + INSTALLER_DIR);
+
+    List<String> violations = new ArrayList<>();
+    for (Path script : scripts) {
+      String text = Files.readString(script, StandardCharsets.UTF_8);
+      Matcher m = XML_COMMENT.matcher(text);
+      while (m.find()) {
+        String inner = m.group(1);
+        if (inner.contains("--")) {
+          int line = 1 + (int) text.substring(0, m.start()).chars().filter(c -> c == '\n').count();
+          violations.add(script.getFileName() + ":" + line + " comment contains \"--\"");
+        }
+      }
+    }
+
+    assertTrue(
+        violations.isEmpty(),
+        () ->
+            "XML comments must not contain \"--\" (XML 1.0 §2.5). Reword CLI flag mentions"
+                + " (e.g. demo-sites flag instead of double-dash demo-sites). Offenders:\n"
+                + String.join("\n", violations));
+  }
+
+  @Test
+  @DisplayName("Installer Ant XML files parse as well-formed XML (#2290)")
+  void installerAntXmlIsWellFormed()
+      throws IOException, ParserConfigurationException, SAXException {
+    DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+    factory.setNamespaceAware(false);
+    factory.setValidating(false);
+    // Do not expand external entities; these are local packaging scripts.
+    factory.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+    factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+    factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+    DocumentBuilder builder = factory.newDocumentBuilder();
+
+    List<Path> scripts = listTopLevelInstallerXml();
+    assertFalse(scripts.isEmpty(), "expected Installer Ant XML under " + INSTALLER_DIR);
+
+    for (Path script : scripts) {
+      try (var in = Files.newInputStream(script)) {
+        builder.parse(in);
+      } catch (SAXException e) {
+        fail("Installer Ant XML is not well-formed: " + script + " — " + e.getMessage(), e);
+      }
+    }
+  }
+
+  /** Top-level Ant scripts only (not data/*.xml table dumps). */
+  private static List<Path> listTopLevelInstallerXml() throws IOException {
+    assertTrue(
+        Files.isDirectory(INSTALLER_DIR),
+        () -> "missing Installer dir: " + INSTALLER_DIR.toAbsolutePath().normalize());
+    try (Stream<Path> stream = Files.list(INSTALLER_DIR)) {
+      return stream
+          .filter(Files::isRegularFile)
+          .filter(p -> p.getFileName().toString().endsWith(".xml"))
+          .sorted()
+          .toList();
+    }
+  }
+}

